### PR TITLE
SE-824-Add-test-and-documentation-for-IProtocolERC20UMin

### PR DESCRIPTION
**Pull request by Palmerg4**
Added tests for a minimal upgradeabale protocol ERC20 implementation.

Quick question for clarity, the IProtocolTokenMin and IProtocolERC20UMin interfaces are essentially the same interface. IProtocolTokenMin has documentation already that would be very similar to IProtocolERC20UMin documentation.

How should I go forward? I can add the documentation for IProtocolERC20UMin if needed, but I think we may just want to remove this interface and continue with IProtocolTokenMin. Thoughts?

### CODE REVIEW

**Comment by ShaneDuncan602** [comment]
> Added tests for a minimal upgradeabale protocol ERC20 implementation.
> 
> Quick question for clarity, the IProtocolTokenMin and IProtocolERC20UMin interfaces are essentially the same interface. IProtocolTokenMin has documentation already that would be very similar to IProtocolERC20UMin documentation.
> 
> How should I go forward? I can add the documentation for IProtocolERC20UMin if needed, but I think we may just want to remove this interface and continue with IProtocolTokenMin. Thoughts?

IMO, we should remove the IProtocolTokenMin and remove the U from IProtocolERC20UMin. I know the hope with IProtocolTokenMin was to provide a simple interface for both ERC20 and ERC721 but I don't think that will ever be something we push. ERC721 has the added dependency of Enumerable so they will always need to have separate interfaces.

**Comment by TJ-Everett** [comment]
> > Added tests for a minimal upgradeabale protocol ERC20 implementation.
> > Quick question for clarity, the IProtocolTokenMin and IProtocolERC20UMin interfaces are essentially the same interface. IProtocolTokenMin has documentation already that would be very similar to IProtocolERC20UMin documentation.
> > How should I go forward? I can add the documentation for IProtocolERC20UMin if needed, but I think we may just want to remove this interface and continue with IProtocolTokenMin. Thoughts?
> 
> IMO, we should remove the IProtocolTokenMin and remove the U from IProtocolERC20UMin. I know the hope with IProtocolTokenMin was to provide a simple interface for both ERC20 and ERC721 but I don't think that will ever be something we push. ERC721 has the added dependency of Enumerable so they will always need to have separate interfaces.

I agree Shane, 
One counterpoint is would this be of any use for other token standards in the future? Such as 1155. Or should we just create new minimal interfaces for those as we build them out?

**Review by TJ-Everett — APPROVED** [review]
Looks great to me Gordon! Approved

**Comment by ShaneDuncan602** [comment]
> > > Added tests for a minimal upgradeabale protocol ERC20 implementation.
> > > Quick question for clarity, the IProtocolTokenMin and IProtocolERC20UMin interfaces are essentially the same interface. IProtocolTokenMin has documentation already that would be very similar to IProtocolERC20UMin documentation.
> > > How should I go forward? I can add the documentation for IProtocolERC20UMin if needed, but I think we may just want to remove this interface and continue with IProtocolTokenMin. Thoughts?
> > 
> > 
> > IMO, we should remove the IProtocolTokenMin and remove the U from IProtocolERC20UMin. I know the hope with IProtocolTokenMin was to provide a simple interface for both ERC20 and ERC721 but I don't think that will ever be something we push. ERC721 has the added dependency of Enumerable so they will always need to have separate interfaces.
> 
> I agree Shane, One counterpoint is would this be of any use for other token standards in the future? Such as 1155. Or should we just create new minimal interfaces for those as we build them out?

I think we would create a new interface for any new supported standard. It makes it simple for the implementing devs to understand and gives us flexibility if one changes.

**Review by ShaneDuncan602 — APPROVED** [review]
Awesome!

File: docs/userGuides/architecture/client/PROTOCOL_ERC20_TOKEN_MINIMAL_IMPLEMENTATION.md [hunk 1]
```diff
@@ -1,12 +1,12 @@
-# Protocol Token Minimal Implementation
+# Protocol ERC20 Token Minimal Implementation
 
 ## Purpose
 
-The [IProtocolTokenMin](../../../../src/client/token/IProtocolTokenMin.sol) interface contains the minimal set of functions that must be implemented in order for a token (ERC20 or ERC721) to be compatible with the protocol.
+The [IProtocolERC20Min](src/client/token/ERC20/IProtocolERC20Min.sol) interface contains the minimal set of functions that must be implemented in order for an ERC20 token to be compatible with the protocol.
```
Comment (VoR0220) [inline docs/userGuides/architecture/client/PROTOCOL_ERC20_TOKEN_MINIMAL_IMPLEMENTATION.md]: broken link

File: docs/userGuides/architecture/client/PROTOCOL_ERC721_TOKEN_MINIMAL_IMPLEMENTATION.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Protocol ERC721 Token Minimal Implementation
+
+## Purpose
+
+The [IProtocolER721Min](src/client/token/ERC721/IProtocolERC721Min.sol) interface contains the minimal set of functions that must be implemented in order for an ERC721 token to be compatible with the protocol.
```
Comment (VoR0220) [inline docs/userGuides/architecture/client/PROTOCOL_ERC721_TOKEN_MINIMAL_IMPLEMENTATION.md]: broken link

File: src/client/token/ERC20/upgradeable/ProtocolERC20UMin.sol [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import "src/client/token/IProtocolTokenHandler.sol";
+import "src/client/token/ProtocolTokenCommonU.sol";
+
+
+/**
+ * @title ERC20 Upgradeable Minimal Protocol Contract
+ * @author @ShaneDuncan602, @oscarsernarosero, @TJ-Everett, @Palmerg4
+ * @notice This is the base contract for all protocol ERC721Upgradeable Minimals. 
+ */
+contract ProtocolERC20UMin is Initializable, ERC20Upgradeable, ProtocolTokenCommonU, ReentrancyGuard, IProtocolERC20Min{
```
Comment (VoR0220) [inline src/client/token/ERC20/upgradeable/ProtocolERC20UMin.sol:16]: ++

File: test/client/token/ERC20/integration/ApplicationERC20UMin.t.sol [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+    }
+
+    function testERC20_ApplicationERC20U_Mint() public endWithStopPrank {
+        switchToAppAdministrator();
+        /// Owner Mints new coins
+        ApplicationERC20UMinUpgAdminMint(address(applicationCoinProxy)).mint(appAdministrator, 100);
+        console.log(ApplicationERC20UMinUpgAdminMint(address(applicationCoinProxy)).balanceOf(appAdministrator));
+        /// Owner Mints a second new coins
+        ApplicationERC20UMinUpgAdminMint(address(applicationCoinProxy)).mint(appAdministrator, 200);
+        console.log(ApplicationERC20UMinUpgAdminMint(address(applicationCoinProxy)).balanceOf(appAdministrator));
```
Comment (VoR0220) [inline test/client/token/ERC20/integration/ApplicationERC20UMin.t.sol]: Maybe drop the console.logs here?

File: test/client/token/ERC20/integration/ApplicationERC20UMin.t.sol [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+        switchToAppAdministrator();
+        applicationAppManager2.confirmAppManager(address(applicationCoinProxy));
+        /// test to ensure it still works
+        ApplicationERC20UMinUpgAdminMint(address(applicationCoinProxy)).mint(appAdministrator, 100);
+        switchToAppAdministrator();
+        ApplicationERC20UMinUpgAdminMint(address(applicationCoinProxy)).transfer(user, 1);
+        assertEq(ApplicationERC20UMinUpgAdminMint(address(applicationCoinProxy)).balanceOf(appAdministrator), 99);
+        assertEq(ApplicationERC20UMinUpgAdminMint(address(applicationCoinProxy)).balanceOf(user), 1);
+    }
+}
```
Comment (VoR0220) [inline test/client/token/ERC20/integration/ApplicationERC20UMin.t.sol:537]: Tests are clean af. Nice work.

**Comment by VoR0220** [comment]
Need to fix the links and maybe consider dropping the console.logs. Other than that this lgtm.

**Pull request merged**